File: solana_toolbox_idl/src/toolbox_idl_type_full_serialize.rs
use std::cmp::max;
use std::str::FromStr;

use anyhow::anyhow;
use anyhow::Context;
use anyhow::Result;
use serde_json::Value;
use solana_sdk::pubkey::Pubkey;
use solana_toolbox_endpoint::ToolboxEndpoint;

use crate::toolbox_idl_type_full::ToolboxIdlTypeFull;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullEnumVariant;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFields;
use crate::toolbox_idl_type_prefix::ToolboxIdlTypePrefix;
use crate::toolbox_idl_type_primitive::ToolboxIdlTypePrimitive;
use crate::toolbox_idl_utils::idl_as_array_or_else;
use crate::toolbox_idl_utils::idl_as_bool_or_else;
use crate::toolbox_idl_utils::idl_as_bytes_or_else;
use crate::toolbox_idl_utils::idl_as_f64_or_else;
use crate::toolbox_idl_utils::idl_as_i64_or_else;
use crate::toolbox_idl_utils::idl_as_object_or_else;
use crate::toolbox_idl_utils::idl_as_str_or_else;
use crate::toolbox_idl_utils::idl_as_u64_or_else;
use crate::toolbox_idl_utils::idl_object_get_key_as_str;
use crate::toolbox_idl_utils::idl_object_get_key_as_u64;
use crate::toolbox_idl_utils::idl_object_get_key_or_else;

impl ToolboxIdlTypeFull {
    pub fn try_serialize(
        &self,
        value: &Value,
        data: &mut Vec<u8>,
        // TODO (FAR) - Config object for allowing numbers as string
        // TODO (FAR) - Config object for pubkey hashmap and prefixes and existing
        // TODO (FAR) - Config object for custom serializers ?
        deserializable: bool,
    ) -> Result<()> {
        match self {
            ToolboxIdlTypeFull::Typedef { name, content, .. } => {
                ToolboxIdlTypeFull::try_serialize(
                    content,
                    value,
                    data,
                    deserializable,
                )
                .with_context(|| format!("Serialize Typedef, name: {}", name))
            },
            ToolboxIdlTypeFull::Option {
                prefix, content, ..
            } => ToolboxIdlTypeFull::try_serialize_option(
                prefix,
                content,
                value,
                data,
                deserializable,
            ),
            ToolboxIdlTypeFull::Vec { prefix, items, .. } => {
                ToolboxIdlTypeFull::try_serialize_vec(
                    prefix,
                    items,
                    value,
                    data,
                    deserializable,
                )
            },
            ToolboxIdlTypeFull::Array { items, length } => {
                ToolboxIdlTypeFull::try_serialize_array(
                    items,
                    length,
                    value,
                    data,
                    deserializable,
                )
            },
            ToolboxIdlTypeFull::String { prefix, .. } => {
                ToolboxIdlTypeFull::try_serialize_string(
                    prefix,
                    value,
                    data,
                    deserializable,
                )
            },
            ToolboxIdlTypeFull::Struct { fields, .. } => {
                ToolboxIdlTypeFull::try_serialize_struct(
                    fields,
                    value,
                    data,
                    deserializable,
                )
            },
            ToolboxIdlTypeFull::Enum {
                prefix, variants, ..
            } => ToolboxIdlTypeFull::try_serialize_enum(
                prefix,
                variants,
                value,
                data,
                deserializable,
            ),
            ToolboxIdlTypeFull::Padded {
                before,
                min_size,
                after,
                content,
            } => ToolboxIdlTypeFull::try_serialize_padded(
                before,
                min_size,
                after,
                content,
                value,
                data,
                deserializable,
            ),
            ToolboxIdlTypeFull::Const { literal } => {
                Err(anyhow!("Can't use a const literal directly: {}", literal))
            },
            ToolboxIdlTypeFull::Primitive { primitive } => {
                primitive.try_serialize(value, data)
            },
        }
    }

    fn try_serialize_option(
        option_prefix: &ToolboxIdlTypePrefix,
        option_content: &ToolboxIdlTypeFull,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        if value.is_null() {
            option_prefix.try_serialize(0, data)?;
        } else {
            option_prefix.try_serialize(1, data)?;
            option_content.try_serialize(value, data, deserializable)?;
        }
        Ok(())
    }

    fn try_serialize_vec(
        vec_prefix: &ToolboxIdlTypePrefix,
        vec_items: &ToolboxIdlTypeFull,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        if vec_items.is_primitive(&ToolboxIdlTypePrimitive::U8) {
            let bytes = try_read_value_to_bytes(value)?;
            if deserializable {
                vec_prefix.try_serialize(u64::try_from(bytes.len())?, data)?;
            }
            data.extend_from_slice(&bytes);
            return Ok(());
        }
        let values = idl_as_array_or_else(value)?;
        if deserializable {
            vec_prefix.try_serialize(u64::try_from(values.len())?, data)?;
        }
        for (index, value_item) in values.iter().enumerate() {
            vec_items
                .try_serialize(value_item, data, deserializable)
                .with_context(|| format!("Serialize Vec Item: {}", index))?;
        }
        Ok(())
    }

    fn try_serialize_array(
        array_items: &ToolboxIdlTypeFull,
        array_length: &usize,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        if array_items.is_primitive(&ToolboxIdlTypePrimitive::U8) {
            let bytes = try_read_value_to_bytes(value)?;
            if bytes.len() != *array_length {
                return Err(anyhow!(
                    "value byte array is not the correct size: expected {} bytes, found {} bytes",
                    array_length,
                    bytes.len()
                ));
            }
            data.extend_from_slice(&bytes);
            return Ok(());
        }
        let values = idl_as_array_or_else(value)?;
        if values.len() != *array_length {
            return Err(anyhow!(
                "value array is not the correct size: expected {} items, found {} items",
                array_length,
                values.len()
            ));
        }
        for (index, value_item) in values.iter().enumerate() {
            array_items
                .try_serialize(value_item, data, deserializable)
                .with_context(|| format!("Serialize Array Item: {}", index))?;
        }
        Ok(())
    }

    fn try_serialize_string(
        string_prefix: &ToolboxIdlTypePrefix,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        let value_str = idl_as_str_or_else(value)?;
        if deserializable {
            string_prefix
                .try_serialize(u64::try_from(value_str.len())?, data)?;
        }
        data.extend_from_slice(value_str.as_bytes());
        Ok(())
    }

    fn try_serialize_struct(
        struct_fields: &ToolboxIdlTypeFullFields,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        struct_fields.try_serialize(value, data, deserializable)
    }

    fn try_serialize_enum(
        enum_prefix: &ToolboxIdlTypePrefix,
        enum_variants: &[ToolboxIdlTypeFullEnumVariant],
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        if let Some(value_number) = value.as_u64() {
            for enum_variant in enum_variants {
                if enum_variant.code == value_number {
                    enum_prefix.try_serialize(enum_variant.code, data)?;
                    return ToolboxIdlTypeFull::try_serialize_enum_variant(
                        enum_variant,
                        &Value::Null,
                        data,
                        deserializable,
                    );
                }
            }
            return Err(anyhow!(
                "Could not find enum variant with code: {}",
                value_number
            ));
        }
        if let Some(value_string) = value.as_str() {
            for enum_variant in enum_variants {
                if enum_variant.name == value_string {
                    enum_prefix.try_serialize(enum_variant.code, data)?;
                    return ToolboxIdlTypeFull::try_serialize_enum_variant(
                        enum_variant,
                        &Value::Null,
                        data,
                        deserializable,
                    );
                }
            }
            return Err(anyhow!(
                "Could not find enum variant with name: {}",
                value_string
            ));
        }
        if let Some(value_object) = value.as_object() {
            for enum_variant in enum_variants {
                if let Some(enum_value) = value_object.get(&enum_variant.name) {
                    enum_prefix.try_serialize(enum_variant.code, data)?;
                    return ToolboxIdlTypeFull::try_serialize_enum_variant(
                        enum_variant,
                        enum_value,
                        data,
                        deserializable,
                    );
                }
            }
            return Err(anyhow!("Could not guess enum from object keys"));
        }
        Err(anyhow!(
            "Expected enum value to be: number/string or object"
        ))
    }

    fn try_serialize_enum_variant(
        enum_variant: &ToolboxIdlTypeFullEnumVariant,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        enum_variant
            .fields
            .try_serialize(value, data, deserializable)
            .with_context(|| {
                format!("Serialize Enum Variant: {}", enum_variant.name)
            })
    }

    fn try_serialize_padded(
        padded_before: &usize,
        padded_min_size: &usize,
        padded_after: &usize,
        padded_content: &ToolboxIdlTypeFull,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        let data_offset_before = data.len() + *padded_before;
        while data.len() < data_offset_before {
            data.push(0);
        }
        padded_content.try_serialize(value, data, deserializable)?;
        let data_content_size = data.len() - data_offset_before;
        let data_offset_after = data_offset_before
            + max(*padded_min_size, data_content_size)
            + *padded_after;
        while data.len() < data_offset_after {
            data.push(0);
        }
        Ok(())
    }
}

impl ToolboxIdlTypeFullFields {
    pub fn try_serialize(
        &self,
        value: &Value,
        data: &mut Vec<u8>,
        deserializable: bool,
    ) -> Result<()> {
        if self.len() == 0 {
            return Ok(());
        }
        match self {
            ToolboxIdlTypeFullFields::Named(fields) => {
                let value = idl_as_object_or_else(value)?;
                for field in fields {
                    let value_field =
                        idl_object_get_key_or_else(value, &field.name)?;
                    field
                        .content
                        .try_serialize(value_field, data, deserializable)
                        .with_context(|| {
                            format!("Serialize Field: {}", field.name)
                        })?;
                }
            },
            ToolboxIdlTypeFullFields::Unnamed(fields) => {
                let values = idl_as_array_or_else(value)?;
                if values.len() != fields.len() {
                    return Err(anyhow!("Wrong number of unnamed fields, expected: {}, found: {}", fields.len(), values.len()));
                }
                for field in fields {
                    let value_field = &values[field.position];
                    field
                        .content
                        .try_serialize(value_field, data, deserializable)
                        .with_context(|| {
                            format!("Serialize Field: {}", field.position)
                        })?;
                }
            },
        }
        Ok(())
    }
}

impl ToolboxIdlTypePrefix {
    pub fn try_serialize(&self, value: u64, data: &mut Vec<u8>) -> Result<()> {
        match self {
            ToolboxIdlTypePrefix::U8 => {
                data.extend_from_slice(&u8::try_from(value)?.to_le_bytes())
            },
            ToolboxIdlTypePrefix::U16 => {
                data.extend_from_slice(&u16::try_from(value)?.to_le_bytes())
            },
            ToolboxIdlTypePrefix::U32 => {
                data.extend_from_slice(&u32::try_from(value)?.to_le_bytes())
            },
            ToolboxIdlTypePrefix::U64 => {
                data.extend_from_slice(&value.to_le_bytes())
            },
            ToolboxIdlTypePrefix::U128 => {
                data.extend_from_slice(&u128::from(value).to_le_bytes())
            },
        }
        Ok(())
    }
}

impl ToolboxIdlTypePrimitive {
    pub fn try_serialize(
        self: &ToolboxIdlTypePrimitive,
        value: &Value,
        data: &mut Vec<u8>,
    ) -> Result<()> {
        match self {
            ToolboxIdlTypePrimitive::U8 => {
                let value_integer = idl_as_u64_or_else(value)?;
                let value_typed = u8::try_from(value_integer)?;
                data.extend_from_slice(&value_typed.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::U16 => {
                let value_integer = idl_as_u64_or_else(value)?;
                let value_typed = u16::try_from(value_integer)?;
                data.extend_from_slice(&value_typed.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::U32 => {
                let value_integer = idl_as_u64_or_else(value)?;
                let value_typed = u32::try_from(value_integer)?;
                data.extend_from_slice(&value_typed.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::U64 => {
                let value_integer = idl_as_u64_or_else(value)?;
                data.extend_from_slice(&value_integer.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::U128 => {
                let value_integer = u128::from(idl_as_u64_or_else(value)?);
                data.extend_from_slice(&value_integer.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::I8 => {
                let value_integer = idl_as_i64_or_else(value)?;
                let value_typed = i8::try_from(value_integer)?;
                data.extend_from_slice(&value_typed.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::I16 => {
                let value_integer = idl_as_i64_or_else(value)?;
                let value_typed = i16::try_from(value_integer)?;
                data.extend_from_slice(&value_typed.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::I32 => {
                let value_integer = idl_as_i64_or_else(value)?;
                let value_typed = i32::try_from(value_integer)?;
                data.extend_from_slice(&value_typed.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::I64 => {
                let value_integer = idl_as_i64_or_else(value)?;
                data.extend_from_slice(&value_integer.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::I128 => {
                let value_integer = i128::from(idl_as_i64_or_else(value)?);
                data.extend_from_slice(&value_integer.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::F32 => {
                let value_floating = idl_as_f64_or_else(value)? as f32;
                data.extend_from_slice(&value_floating.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::F64 => {
                let value_floating = idl_as_f64_or_else(value)?;
                data.extend_from_slice(&value_floating.to_le_bytes());
            },
            ToolboxIdlTypePrimitive::Bool => {
                let value_boolean = idl_as_bool_or_else(value)?;
                data.push(if value_boolean { 1 } else { 0 });
            },
            ToolboxIdlTypePrimitive::Pubkey => {
                let value_str = idl_as_str_or_else(value)?;
                let value_pubkey = Pubkey::from_str(value_str)?;
                data.extend_from_slice(&value_pubkey.to_bytes());
            },
        };
        Ok(())
    }
}

fn try_read_value_to_bytes(value: &Value) -> Result<Vec<u8>> {
    if let Some(value_array) = value.as_array() {
        return idl_as_bytes_or_else(value_array);
    }
    if let Some(value_object) = value.as_object() {
        if let Some(data) = idl_object_get_key_as_str(value_object, "base16") {
            return ToolboxEndpoint::sanitize_and_decode_base16(data);
        }
        if let Some(data) = idl_object_get_key_as_str(value_object, "base58") {
            return ToolboxEndpoint::sanitize_and_decode_base58(data);
        }
        if let Some(data) = idl_object_get_key_as_str(value_object, "base64") {
            return ToolboxEndpoint::sanitize_and_decode_base64(data);
        }
        if let Some(data) = idl_object_get_key_as_str(value_object, "utf8") {
            return Ok(data.as_bytes().to_vec());
        }
        if let Some(data) = idl_object_get_key_as_u64(value_object, "zeroes") {
            return Ok(vec![0; usize::try_from(data)?]);
        }
    }
    Err(anyhow!("Could not read bytes, expected an array/object"))
}

File: solana_toolbox_web3/src/ToolboxIdlTypeFull.serialize.ts
import { PublicKey } from '@solana/web3.js';
import {
  ToolboxIdlTypeFull,
  ToolboxIdlTypeFullArray,
  ToolboxIdlTypeFullConst,
  ToolboxIdlTypeFullEnum,
  ToolboxIdlTypeFullEnumVariant,
  ToolboxIdlTypeFullFieldNamed,
  ToolboxIdlTypeFullFields,
  ToolboxIdlTypeFullFieldUnnamed,
  ToolboxIdlTypeFullOption,
  ToolboxIdlTypeFullPadded,
  ToolboxIdlTypeFullString,
  ToolboxIdlTypeFullStruct,
  ToolboxIdlTypeFullTypedef,
  ToolboxIdlTypeFullVec,
} from './ToolboxIdlTypeFull';
import { ToolboxIdlTypePrimitive } from './ToolboxIdlTypePrimitive';
import { ToolboxUtils } from './ToolboxUtils';
import { ToolboxIdlTypePrefix } from './ToolboxIdlTypePrefix';

export function serialize(
  typeFull: ToolboxIdlTypeFull,
  value: any,
  data: Buffer[],
  prefixed: boolean,
) {
  typeFull.traverse(serializeVisitor, value, data, prefixed);
}

let serializeVisitor = {
  typedef: (
    self: ToolboxIdlTypeFullTypedef,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    ToolboxUtils.withContext(() => {
      return serialize(self.content, value, data, prefixed);
    }, `Serialize: Typedef: ${self.name}`);
  },
  option: (
    self: ToolboxIdlTypeFullOption,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    if (value === null) {
      serializePrefix(self.prefix, 0, data);
      return;
    }
    serializePrefix(self.prefix, 1, data);
    serialize(self.content, value, data, prefixed);
  },
  vec: (
    self: ToolboxIdlTypeFullVec,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    let array = ToolboxUtils.expectArray(value);
    if (prefixed) {
      serializePrefix(self.prefix, array.length, data);
    }
    for (let item of array) {
      serialize(self.items, item, data, prefixed);
    }
  },
  array: (
    self: ToolboxIdlTypeFullArray,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    let array = ToolboxUtils.expectArray(value);
    if (array.length != self.length) {
      throw new Error(`Expected an array of size: ${self.length}`);
    }
    for (let item of array) {
      serialize(self.items, item, data, prefixed);
    }
  },
  string: (
    self: ToolboxIdlTypeFullString,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    let string = ToolboxUtils.expectString(value);
    if (prefixed) {
      serializePrefix(self.prefix, string.length, data);
    }
    data.push(Buffer.from(string, 'utf8'));
  },
  struct: (
    self: ToolboxIdlTypeFullStruct,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    serializeFields(self.fields, value, data, prefixed);
  },
  enum: (
    self: ToolboxIdlTypeFullEnum,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    function serializeEnumVariant(
      variant: ToolboxIdlTypeFullEnumVariant,
      value: any,
    ) {
      ToolboxUtils.withContext(() => {
        serializePrefix(self.prefix, variant.code, data);
        serializeFields(variant.fields, value, data, prefixed);
      }, `Serialize: Enum Variant: ${variant.name}`);
    }
    if (ToolboxUtils.isNumber(value)) {
      for (let variant of self.variants) {
        if (variant.code == value) {
          return serializeEnumVariant(variant, null);
        }
      }
      throw new Error(`Could not find enum variant with code: ${value}`);
    }
    if (ToolboxUtils.isString(value)) {
      for (let variant of self.variants) {
        if (variant.name == value) {
          return serializeEnumVariant(variant, null);
        }
      }
      throw new Error(`Could not find enum variant with name: ${value}`);
    }
    if (ToolboxUtils.isObject(value)) {
      for (let variant of self.variants) {
        if (value.hasOwnProperty(variant.name)) {
          return serializeEnumVariant(variant, value[variant.name]);
        }
      }
      throw new Error('Could not guess enum variant from object key');
    }
    throw new Error('Expected enum value to be: number/string or object');
  },
  padded: (
    self: ToolboxIdlTypeFullPadded,
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    if (self.before) {
      data.push(Buffer.alloc(self.before));
    }
    let contentData: Buffer[] = [];
    serialize(self.content, value, contentData, prefixed);
    for (let contentBuffer of contentData) {
      data.push(contentBuffer);
    }
    let contentSize = contentData.reduce((size, contentBuffer) => {
      return size + contentBuffer.length;
    }, 0);
    if (self.minSize > contentSize) {
      data.push(Buffer.alloc(self.minSize - contentSize));
    }
    if (self.after) {
      data.push(Buffer.alloc(self.after));
    }
  },
  const: (
    _self: ToolboxIdlTypeFullConst,
    _value: any,
    _data: Buffer[],
    _prefixed: boolean,
  ) => {
    throw new Error('Cannot serialize a const type');
  },
  primitive: (
    self: ToolboxIdlTypePrimitive,
    value: any,
    data: Buffer[],
    _prefixed: boolean,
  ) => {
    serializePrimitive(self, value, data);
  },
};

export function serializeFields(
  typeFullFields: ToolboxIdlTypeFullFields,
  value: any,
  data: Buffer[],
  prefixed: boolean,
) {
  typeFullFields.traverse(serializeFieldsVisitor, value, data, prefixed);
}

let serializeFieldsVisitor = {
  named: (
    self: ToolboxIdlTypeFullFieldNamed[],
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    if (self.length <= 0) {
      return;
    }
    ToolboxUtils.expectObject(value);
    for (let field of self) {
      ToolboxUtils.withContext(() => {
        serialize(field.content, value[field.name], data, prefixed);
      }, `Serialize: Field: ${field.name}`);
    }
  },
  unnamed: (
    self: ToolboxIdlTypeFullFieldUnnamed[],
    value: any,
    data: Buffer[],
    prefixed: boolean,
  ) => {
    if (self.length <= 0) {
      return;
    }
    ToolboxUtils.expectArray(value);
    for (let field of self) {
      ToolboxUtils.withContext(() => {
        serialize(field.content, value[field.position], data, prefixed);
      }, `Serialize: Field: ${field.position}`);
    }
  },
};

export function serializePrefix(
  prefix: ToolboxIdlTypePrefix,
  value: number,
  data: Buffer[],
) {
  let buffer = Buffer.alloc(prefix.size);
  prefix.traverse(serializePrefixVisitor, buffer, value);
  data.push(buffer);
}

let serializePrefixVisitor = {
  u8: (buffer: Buffer, value: any) => {
    buffer.writeUInt8(ToolboxUtils.expectNumber(value));
  },
  u16: (buffer: Buffer, value: any) => {
    buffer.writeUInt16LE(ToolboxUtils.expectNumber(value));
  },
  u32: (buffer: Buffer, value: any) => {
    buffer.writeUInt32LE(ToolboxUtils.expectNumber(value));
  },
  u64: (buffer: Buffer, value: any) => {
    buffer.writeBigUInt64LE(BigInt(ToolboxUtils.expectNumber(value)));
  },
  u128: (buffer: Buffer, value: any) => {
    buffer.writeBigUInt64LE(BigInt(ToolboxUtils.expectNumber(value)));
  },
};

export function serializePrimitive(
  primitive: ToolboxIdlTypePrimitive,
  value: any,
  data: Buffer[],
) {
  let buffer = Buffer.alloc(primitive.size);
  primitive.traverse(serializePrimitiveVisitor, buffer, value);
  data.push(buffer);
}

let serializePrimitiveVisitor = {
  u8: (buffer: Buffer, value: any) => {
    buffer.writeUInt8(ToolboxUtils.expectNumber(value));
  },
  u16: (buffer: Buffer, value: any) => {
    buffer.writeUInt16LE(ToolboxUtils.expectNumber(value));
  },
  u32: (buffer: Buffer, value: any) => {
    buffer.writeUInt32LE(ToolboxUtils.expectNumber(value));
  },
  u64: (buffer: Buffer, value: any) => {
    buffer.writeBigUInt64LE(ToolboxUtils.expectBigInt(value));
  },
  u128: (buffer: Buffer, value: any) => {
    let num = ToolboxUtils.expectBigInt(value);
    let low = num & 0xffffffffffffffffn;
    let high = (num >> 64n) & 0xffffffffffffffffn;
    buffer.writeBigUInt64LE(low, 0);
    buffer.writeBigUInt64LE(high, 8);
  },
  i8: (buffer: Buffer, value: any) => {
    buffer.writeInt8(ToolboxUtils.expectNumber(value));
  },
  i16: (buffer: Buffer, value: any) => {
    buffer.writeInt16LE(ToolboxUtils.expectNumber(value));
  },
  i32: (buffer: Buffer, value: any) => {
    buffer.writeInt32LE(ToolboxUtils.expectNumber(value));
  },
  i64: (buffer: Buffer, value: any) => {
    buffer.writeBigInt64LE(ToolboxUtils.expectBigInt(value));
  },
  i128: (buffer: Buffer, value: any) => {
    let num = ToolboxUtils.expectBigInt(value);
    let low = BigInt.asIntN(64, num);
    let high = BigInt.asIntN(64, num >> 64n);
    buffer.writeBigInt64LE(low, 0);
    buffer.writeBigInt64LE(high, 8);
  },
  f32: (buffer: Buffer, value: any) => {
    buffer.writeFloatLE(ToolboxUtils.expectNumber(value));
  },
  f64: (buffer: Buffer, value: any) => {
    buffer.writeDoubleLE(ToolboxUtils.expectNumber(value));
  },
  bool: (buffer: Buffer, value: any) => {
    if (ToolboxUtils.expectBoolean(value)) {
      buffer.writeUInt8(1);
    } else {
      buffer.writeUInt8(0);
    }
  },
  pubkey: (buffer: Buffer, value: any) => {
    buffer.set(new PublicKey(ToolboxUtils.expectString(value)).toBuffer());
  },
};
